modules/experimental-layers/src/contour-layer/marching-squares.js

```diff
@@ -32,57 +32,66 @@ const CODE_OFFSET_MAP = {
 };
 
 // Returns marching square code for given cell
+/* eslint-disable complexity */
 export function getCode(params) {
   // Assumptions
   // Origin is on bottom-left , and X increase to right, Y to top
   // When processing one cell, we process 4 cells, by extending row to top and on column to right
   // to create a 2X2 cell grid
 
-  const {cellWeights, thresholdValue, cellIndex, gridSize} = params;
-
-  const numRows = gridSize[1];
-  const numCols = gridSize[0];
-
-  // TODO: duplicate top row and right column
-  // We shouldn't process the right column
-  assert((cellIndex + 1) % numCols);
-  // We shouldn't process the topmost row
-  assert(cellIndex + 1 < (numRows - 1) * numCols);
-
-  const top = cellWeights[cellIndex + numCols] - thresholdValue >= 0 ? 1 : 0;
-  const topRight = cellWeights[cellIndex + numCols + 1] - thresholdValue >= 0 ? 1 : 0;
-  const right = cellWeights[cellIndex + 1] - thresholdValue >= 0 ? 1 : 0;
-  const current = cellWeights[cellIndex] - thresholdValue >= 0 ? 1 : 0;
+  const {cellWeights, thresholdValue, x, y, width, height} = params;
+
+  assert(x >= -1 && x < width);
+  assert(y >= -1 && y < height);
+
+  const leftBoundary = x < 0;
+  const rightBoundary = x >= width - 1;
+  const bottomBoundary = y < 0;
+  const topBoundary = y >= height - 1;
+
+  const top =
+    leftBoundary || topBoundary
+      ? 0
+      : cellWeights[(y + 1) * width + x] - thresholdValue >= 0
+        ? 1
+        : 0;
+  const topRight =
+    rightBoundary || topBoundary
+      ? 0
+      : cellWeights[(y + 1) * width + x + 1] - thresholdValue >= 0
+        ? 1
+        : 0;
+  const right = rightBoundary ? 0 : cellWeights[y * width + x + 1] - thresholdValue >= 0 ? 1 : 0;
+  const current =
+    leftBoundary || bottomBoundary ? 0 : cellWeights[y * width + x] - thresholdValue >= 0 ? 1 : 0;
 
   const code = (top << 3) | (topRight << 2) | (right << 1) | current;
 
   assert(code >= 0 && code < 16);
 
   return code;
 }
+/* eslint-enable complexity */
 
 // Returns intersection vertices for given cellindex
 export function getVertices(params) {
-  const {gridOrigin, cellIndex, cellSize, gridSize, code} = params;
+  const {gridOrigin, cellSize, x, y, code} = params;
 
   const offsets = CODE_OFFSET_MAP[code];
   // Reference vertex is top-right its co-ordinates are stored at index 0(X) and 1(Y)
-  const row = Math.floor(cellIndex / gridSize[0]);
-  const col = cellIndex - row * gridSize[0];
-
   // Move to top-right corner
-  const rX = (col + 1) * cellSize[0];
-  const rY = (row + 1) * cellSize[1];
+  const rX = (x + 1) * cellSize[0];
```

Comment:
Am I correct that this function accept negative values for `x` and then adds `1` here? It saw the loops  that start at `-1` above.

Seems like it could be a "gotcha" - maybe mention in a comment before this function.

Reply:
yes these can have `-1` value, since reference corner is always top-right we add `1` every time. I will add an extra comment.